nixos/nvidia: allow compute workloads

eadwu opened the pull request:
## Description of changes
Have been using some rendition of this in the past, but since I had to rebase might as well create the PR.

Nothing really much, doesn't add anything, reorganizes existing module into two separate sections (driver-only and w/ display).

Personally this is a relatively important setup for hot [un]plugging. 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Gerg-L commented:
I like these changes, 
you should squash your commits
and maybe change the `enable` option to something more verbose

Gerg-L commented:
Also I would use this myself as I dynamically bind and unbind my nvidia GPU and generate my own xorg.conf

eadwu commented:
Updated.

Interesting side-effect I observed is that NVIDIA Offload works regardless of the other offload configuration being implemented or not - at least on Wayland (Hyrpland).

Gerg-L commented:
> Interesting side-effect I observed is that NVIDIA Offload works regardless of the other offload configuration being implemented or not - at least on Wayland (Hyrpland).

All the prime options only do xorg.conf configuration and some startup xrandr commands nothing that effects Wayland

eadwu committed:
hardware/video/nvidia: datacenter -> nvidia

This doesn't really make sense, but that's how it is so just keep datacenter as something separate

eadwu committed:
nvidiaEnabled -> hasNvidiaVideoDriver

eadwu committed:
enabled -> enable

eadwu committed:
hardware/nvidia: compute only

eadwu commented:
Pushed variant I have been running locally for ~1yr.

Useful for Kubernetes & VFIO as well now, i.e.
```
sudo modprobe vfio
sudo modprobe vfio_iommu_type1
sudo modprobe vfio_pci

sudo systemctl stop nvidia-persistenced
echo 0000:XX:00.0 | sudo tee /sys/bus/pci/devices/0000:XX:00.0/driver/unbind
echo 0000:XX:00.1 | sudo tee /sys/bus/pci/devices/0000:XX:00.1/driver/unbind
echo '10de XXXX'  | sudo tee /sys/bus/pci/drivers/vfio-pci/new_id
echo '10de XXXX'  | sudo tee /sys/bus/pci/drivers/vfio-pci/remove_id
echo '10de XXXX'  | sudo tee /sys/bus/pci/drivers/vfio-pci/new_id
echo '10de XXXX'  | sudo tee /sys/bus/pci/drivers/vfio-pci/remove_id

echo 0000:XX:00.0 | sudo tee /sys/bus/pci/devices/0000:XX:00.0/driver/unbind
echo 0000:XX:00.1 | sudo tee /sys/bus/pci/devices/0000:XX:00.1/driver/unbind
echo 0000:XX:00.0 | sudo tee /sys/bus/pci/drivers/nvidia/bind
echo 0000:XX:00.1 | sudo tee /sys/bus/pci/drivers/snd_hda_intel/bind
sudo systemctl start nvidia-persistenced
```

eadwu committed:
...

eadwu committed:
nixfmt

phanirithvij commented:
there's a merge conflict.
